### PR TITLE
HDDS-7971. Support SCM transfer Ratis leadership

### PR DESCRIPTION
**Pull request by Xushaohong**
## What changes were proposed in this pull request?
Support SCM transfer Ratis leadership.
Split From https://issues.apache.org/jira/browse/HDDS-5686

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7971

## How was this patch tested?
CI and Manual Test

image

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java [hunk 1]
```diff
@@ -782,6 +795,89 @@ public ScmInfo getScmInfo() throws IOException {
...
+        RaftPeer curLeader = ((SCMRatisServerImpl) scm.getScmHAManager()
+            .getRatisServer()).getLeader();
+        targetPeerId = group.getPeers()
+            .stream()
+            .filter(a -> !a.equals(curLeader)).findFirst()
+            .map(RaftPeer::getId)
+            .orElseThrow(() -> new IOException("Cannot" +
+                " find a new leader to transfer leadership."));
+      } else {
+        String ratisAddr = convertToRatisAddr(newLeaderId);
```
Comment (szetszwo) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:822]: Use `RaftPeerId.getRaftPeerId(newLeaderId)` to get `RaftPeerId`; see https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMRatisServerImpl.java#L172

**Review by szetszwo — COMMENTED** [review]
@Xushaohong , thanks for working on this!  Please see the comment inlined.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java (hunk 1 above)
Comment (Xushaohong) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:822]: Hi @szetszwo , SCM is different from OM as it uses UUID as raftPeerId, See

```
  public String getScmId() {
    return getScmStorageConfig().getScmId();
  }
```

https://github.com/apache/ozone/blob/e7d57b9f911ca9878f2255cf79d8b23b9c2562f3/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMStorageConfig.java#L90-L98

Comment (szetszwo) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:822]: > ... as it uses UUID as raftPeerId, ...

What is the problem of using UUID?  Too long?

If we want to support passing an address, we should have a `--newLeaderAddress` option.

Comment (Xushaohong) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:822]: > What is the problem of using UUID? Too long?


Hi @szetszwo.
OM uses the OmNodeId in the conf as the **RaftPeerId**, so we can directly find the new leader in the Raft group info by the passed nodeId.
SCM uses the UUID as the **RaftPeerId**, we have no way to know the new leader's UUID, so I added the conversion from ScmNodeId to RatisAddress and then got the **RaftPeerId**.

Comment (Xushaohong) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java]: @szetszwo , We have had scmId in the urrent Role info. The sideeffect is the different shell usage between OM and SCM, but seems not a big problem.  I have updated the code.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**